src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java

```diff
@@ -483,6 +435,32 @@ public void testGetShardCacheAgeEmptyForcesUpdate() {
         verify(mockClient).listShards(any());
     }
 
+    @Test
+    public void testNoDuplicateShardsInPartialFailure() {
+        proxy.setCachedShardMap(null);
+        ListShardsResult firstPage = new ListShardsResult().withShards(shards.subList(0, 2)).withNextToken(NEXT_TOKEN);
+        ListShardsResult lastPage = new ListShardsResult().withShards(shards.subList(2, shards.size())).withNextToken(null);
+        when(mockClient.listShards(any())).thenReturn(firstPage).thenThrow(new RuntimeException("Failed!")) // First call
+                .thenReturn(firstPage).thenReturn(lastPage); // second call
+
+        try {
+            proxy.getShardList();
+        } catch (Exception e) {
+            // Do nothing
+        }
+        assertEquals(shards, proxy.getShardList());
```

Comment:
I think it's better to validate the shard Ids and the order as well here.

Reply:
Discussed offline, this is already checked for.